destination/destination.go

```diff
@@ -0,0 +1,360 @@
+// Copyright © 2022 Meroxa, Inc. and Miquido
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package destination
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sync"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/miquido/conduit-connector-elasticsearch/internal/elasticsearch"
+)
+
+func NewDestination() sdk.Destination {
+	return &Destination{}
+}
+
+type Destination struct {
+	sdk.UnimplementedDestination
+
+	config          Config
+	client          client
+	mutex           sync.Mutex
+	operationsQueue BufferQueue
+}
+
+const (
+	actionInsert  = "insert"
+	actionCreate  = "create"
+	actionCreated = "created"
+	actionUpdate  = "update"
+	actionUpdated = "updated"
+	actionDelete  = "delete"
+	actionDeleted = "deleted"
+)
+
+//go:generate moq -out client_moq_test.go . client
+type client = elasticsearch.Client
+
+// GetClient returns the current Elasticsearch client
+func (d *Destination) GetClient() elasticsearch.Client {
+	return d.client
+}
+
+func (d *Destination) Configure(_ context.Context, cfgRaw map[string]string) (err error) {
+	d.config, err = ParseConfig(cfgRaw)
+
+	return
+}
+
+func (d *Destination) Open(ctx context.Context) (err error) {
+	// Initialize Elasticsearch client
+	d.client, err = elasticsearch.NewClient(d.config.Version, d.config)
+	if err != nil {
+		return fmt.Errorf("connection could not be established: %w", err)
+	}
+
+	// Check the connection
+	if err := d.client.Ping(ctx); err != nil {
+		return fmt.Errorf("connection could not be established: %w", err)
+	}
+
+	// Initialize the buffer
+	d.mutex = sync.Mutex{}
+	d.operationsQueue = make(BufferQueue, 0, d.config.BulkSize)
+
+	return nil
+}
+
+func (d *Destination) WriteAsync(ctx context.Context, record sdk.Record, ackFunc sdk.AckFunc) error {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	d.operationsQueue.Enqueue(&operation{
+		CreatedAt: record.CreatedAt,
+		Record:    record,
+		AckFunc:   ackFunc,
+	})
+
+	if uint64(d.operationsQueue.Len()) >= d.config.BulkSize {
+		if err := d.Flush(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *Destination) Flush(ctx context.Context) error {
+	// check if there are operations in the buffer
+	if d.operationsQueue.Empty() {
+		return nil
+	}
+
+	// Sort operations to ensure the order
+	d.operationsQueue.Sort()
```

Comment:
That's correct, is to ensure the order of events as much as it is possible, but still there may be an edge case when two consecutive events were sent, but the second got earlier, filled the buffer and the change was applied.

I'm afraid there is not much to do about this. If we remove sorting logic, then we can get a much bigger mess where the situation You described happens more often. Increasing the buffer size and sorting minimizes this, on average.
A possible "solution" (workaround, but still not perfect) for that is to implement in the Conduit something similar to TCP/IP packets, where they are numbered and when client gets response no.1 and no.3, still waits for no.2 (e.g. some BatchID attached to Record). But even then:
- There is no guarantee that no.2 will ever come, and a pipeline will be stuck. Timeouts, if added, will become a real pain when this happens frequently.
- There is no guarantee that source connector received data in order. It would have to be implemented by actual source.

The order does not always matter. When Record.Key is not set, then records are always created. This also solves it for certain use cases. For that, I think it would be useful to have a dedicated config "Always create" to disable upsert logic event when Record.Key is available.

Reply:
I agree with that you're saying. Let's add a comment to that line above, briefly explaining why we do that. If we ever start seeing the mentioned edge cases or performance degradation, we can work on it then, it's probably too early to work on it now anyway.

Reply:
I'm curious why are we even concerned about records arriving out of order? Conduit guarantees [message order](https://github.com/ConduitIO/conduit/blob/main/docs/pipeline_semantics.md#messages-are-delivered-in-order), so unless the source produces the records out of order I don't see why we would need to sort them. Even if that is the case, it sounds like a bug that needs to be fixed by the source and not something the destination should address.

Reply:
@donatorsky IMHO the point @lovromazgon made is a good one. Given that the sort is running always (so there's always _some_ performance penalty), but that it's not a 100% effective solution (due to batches) and that another condition is that sources messed something up (because as Lovro said, Conduit itself guarantees message order), it looks like we shouldn't be doing it. What are your thoughts on this?

Reply:
@lovromazgon, @hariso,
Got your points, however, I'm still not so sure of that. Let's work on an example to make sure we are on the same page, and I'm not missing anything. In the Salesforce source connector, I set `Record.CreatedAt` based on the date when the event was issued by the Salesforce, not when it was received by the pipeline:
https://github.com/miquido/conduit-connector-salesforce/blob/feat/source-connector/source/source.go#L130

It may happen that there are two events sent **by Salesforce**:
1st: Sent at _15:30:01_, received _15:30:12_.
2nd: Sent at _15:30:02_, received _15:30:11_.

Will Elasticsearch (or any in fact) destination connector receive events in order of them being received (2nd, then 1st) or them being issued by Salesforce (in other words: by `Record.CreatedAt`) meaning Conduit is already doing buffering+sorting?

Because I think @lovromazgon You are saying that Conduit manages the order of messages received, but not the actual order of messages specified in the payload. And, by this `Sort()`, I'm targetting the second option here.

Reply:
In your example you say that those two events were **sent** at those times. With that, you mean issued by Salesforce?

Reply:
That's correct, issued by the Salesforce, and they have these dates in the payload:
1st: Sent at 15:30:01 with payload `{"created_at":"yyyy-mm-dd 15:30:01"}`, received 15:30:12.
2nd: Sent at 15:30:02 with payload `{"created_at":"yyyy-mm-dd 15:30:02"}`, received 15:30:11.

Reply:
:100: If, for any reason, sorting is needed, then the source connector sounds like a better place for that. Plus, sorting will be done even when it's not needed, which affects performance (even if small) for no reason.